vectorstores/pgvector/engine.go

```diff
@@ -0,0 +1,141 @@
+package pgvector
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/jackc/pgx/v5"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/oauth2/v2"
+	"google.golang.org/api/option"
+)
+
+const Port = 5432 // default PostgreSQL port
+
+type ServiceAccountRetriever interface {
+	GetServiceAccountEmail(ctx context.Context) (string, error)
+}
+type DefaultServiceAccountRetriever struct{}
+
+func (d *DefaultServiceAccountRetriever) GetServiceAccountEmail(ctx context.Context) (string, error) {
+	return getServiceAccountEmail(ctx)
+}
+
+type Column struct {
+	Name     string
+	DataType string
+	Nullable bool
+}
+
+func NewColumn(name, dataType string, nullable bool) (*Column, error) {
+	if name == "" {
+		return nil, errors.New("name should be provided")
+	}
+	if dataType == "" {
+		return nil, errors.New("dataType should be provided")
+	}
+	return &Column{Name: name, DataType: dataType, Nullable: nullable}, nil
+}
+
+type PostgresEngine struct {
+	Host                    string
+	Port                    int
+	Database                string
+	User                    string
+	Password                string
+	IAMAccountEmail         string
+	ServiceAccountRetriever ServiceAccountRetriever
+}
+
+// NewPostgresEngine initializes a new PostgresEngine
+func NewPostgresEngine(host, dbname, user, password, iamAccountEmail string, port int) *PostgresEngine {
+	return &PostgresEngine{
```

Comment:
The engine should support arguments: project id, region, cluster, instance, database, user, password, ip type, and iam account email

Reply:
There are going to be many parameters and some of them optional -- would suggest the function optionals pattern here: https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis